### ISSUE TITLE
BeEquivalentTo option ThrowingOnMissingMembers does not work as expected

### ISSUE THREAD

**Issue by lg2de** [fluentassertions/fluentassertions#1249]
### Description

The option `ThrowingOnMissingMembers` for `BeEquivalentTo` does not work as expected.

### Complete minimal example reproducing the issue

```csharp
class Class1
{
    public string Name { get; set; }
}

class Class2
{
    public string Name { get; set; }
    public string Street { get; set; }
}

[TestMethod]
public void BeEquivalentTo1()
{
    new Class1().Should().BeEquivalentTo(new Class2());
}

[TestMethod]
public void BeEquivalentTo2()
{
    new Class2().Should().BeEquivalentTo(new Class1());
}

[TestMethod]
public void BeEquivalentTo3()
{
    new Class2().Should().BeEquivalentTo(new Class1(), o => o.ThrowingOnMissingMembers());
}
```

### Expected vs. actual behavior:

* `BeEquivalentTo1` should fail because `Street` is missing => OK
* `BeEquivalentTo2` should not fail even `Street` is missing in expectation (default behavior) => OK
* `BeEquivalentTo3` should fail because `Street` is missing in expectation => **NOT OK**

Here is the current documentation of `ThrowingOnMissingMembers`:
`Requires the expectation to have members which are equally named to members on the subject.`

According to this documentation `BeEquivalentTo3` should fail, but did not.

Further I did not found any test refer to `ThrowingOnMissingMembers`.

### Versions

using FA version 5.1.0 on .NET Framework 4.7.2

**Comment by dennisdoomen** [comment]
Actually, the documentation is off. `ThrowingOnMissingMember` should throw if the expectation defines a property that the subject doesn't have.

**Comment by lg2de** [comment]
Mmh, this is confusing to me.
1. The **default behavior** of `BeEquivalentTo` is (according to my experience) to throw if the expectation has a property which is not available in subject or different.
2. `ExcludingMissingMembers`: _Tries to match the members of the subject with equally named members on the expectation. Ignores those members that don't exist on the expectation and previously registered matching rules._
3. `ThrowingOnMissingMembers`: _Requires the subject to have members which are equally named to members on the expectation._

All this sounds very similar. I analyzed the code and found that `ThrowingOnMissingMembers` is the default. Further I think the documentation of `ExcludingMissingMembers` could be wrong too.

I think it would be helpful to extend the document and emphasize the differences.
Here my recommendation:
`ExcludingMissingMembers`: **Requires the subject to be equal in all members defined in the expectation.** Ignores those members that don't exist **in the subject or** on the expectation and previously registered matching rules.
`ThrowingOnMissingMembers`: Requires the subject to have **strictly all** members **equal** to members on the expectation. **Ignores those members that don't exist on the expectation.**

Finally, is there a way to check the equivalence as described in the documentation before?
If not, how would you name it?

**Comment by krajek** [comment]
@lg2de hey, coincidentally I had the same issue a few days ago and asked on the slack channel. After a short discussion, we ended up with initial idea of `ThrowingOnUnexpectedMembers` that would work symmetrically to `ThrowingOnMissingMembers`. I will provide a PR shortly after #1252 and #1250 are merged.

**Comment by lg2de** [comment]
Greate @krajek, please drop me a note on the new PR.
I'll have a look on it.
Maybe you can link this issue to you PR.

**Comment by dennisdoomen** [comment]
> All this sounds very similar. I analyzed the code and found that ThrowingOnMissingMembers is the default. Further I think the documentation of ExcludingMissingMembers could be wrong too.

Yes, you're right about both. In FA 5 we flipped the focus to the expectation. I'm actually fixing some docs as we speak.

**Comment by krajek** [comment]
I was talking about preparing a PR fast and I gave it a shot.

Unfortunately, preparing `ThrowingOnUnexpectedMembers` is not that easy, because of the code (`    IEquivalencyAssertionOptions`, `StructuralEqualityEquivalencyStep`, `IMemberMatchingRule` and others) is "one-sided" meaning there is no symmetry of expectation and subject. The subject is used in a few places with no possible easy substitute.

I will work on that, but it possibly be will be much bigger than "just another option".

In the meantime, this is the off-the-band assertion that I am using for my code:

```
public static class StrictReferenceTypeAssertions
    {
        public static void BeStrictlyEquivalentTo<TExpectation>(
            this ObjectAssertions self,
            TExpectation expectation,
            string because = "",
            params object[] becauseArgs)
        {
            var subjectProperties = self.Subject.GetType().GetProperties();

            var expectedObjectProperties = expectation.GetType().GetProperties()
                .ToDictionary(x => x.Name, x => x);

            foreach (PropertyInfo subjectProperty in subjectProperties)
            {
                Execute.Assertion
                    .BecauseOf(because, becauseArgs)
                    .ForCondition(expectedObjectProperties.ContainsKey(subjectProperty.Name))
                    .FailWith("Required property {0} is not present in expected object.",
                        subjectProperty.Name);
            }

            self.BeEquivalentTo(expectation, because, becauseArgs);
        }
    }
```

**Comment by dennisdoomen** [comment]
If it helps, we _can_ introduce breaking changes on `develop`.

**Comment by jnyrup** [comment]
Just two observations:
* `GetProperties()` includes static properties and does not include interface properties.
* I'm not sure if it's possible, but if `expectation` has multiple properties with same `Name`, `ToDictionary` will throw an exception, whereas using `ToLookup` wouldn't.

**Comment by lg2de** [comment]
Today I tried to implement this feature too.
So far I failed (like @krajek) because all the "selection rules" (for members) are focused on the expectation. For this feature it should be possible to switch the view.
Now I fully understand your comment, krajek...

Hope @dennisdoomen or @jnyrup (with better understanding of the internal structure) will be able to implement the feature or to give required tips.

**Comment by jairbubbles** [comment]
*Hey there, I stumbled on the same issue. Here is a bit of context: I was playing  with [Verify](https://www.nuget.org/packages/Verify). I liked some aspects of it but not the fact that the expected results are store in separated .txt files. After discussing with @meziantou I ended up using FluentAssertions with anonymous types which is a powerful combo! By default in Verify it checks that all properties match the result and that none are missing.*

Wouldn't the "strict" mode consist to test also with `expected` and `result` swapped? With a helper function it's trivial:

```csharp
public static void ShouldBeStrictlyEquivalent<TResult, TExpected>(this TResult result, TExpected expected)
{
    result.Should().BeEquivalentTo(expected);
    expected.Should().BeEquivalentTo(result);
}
```

**Comment by dennisdoomen** [comment]
It probably works, but all the detail reporting `BeEquivalentTo` produces would be off.
